docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,380 @@
+# MeshLoadBalancingStrategy - locality aware
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/2689
+
+## Context and Problem Statement
+Right now we have really simple locality aware load balancing, we only care about zones. Dataplane will prefer sending traffic to local zone with fallback to other zones. We would like to add more sophisticated mechanism of locality aware load balancing. We would like to allow user to specify load balancing rules based on tags. For example, big cluster is deployed in multiple AZ’s, and we would like to prefer staying in a single AZ to reduce traffic cost.
+
+## Considered Options
+
+## Decision Outcome
+
+## Solution
+As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone (disable all cross-zone traffic)
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+8. Load balance equally within a group of zones
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
+      crossZone: # (2)
+        rules: # (3)
+          - groups: # (4)
+            - name: "group1" # (5)
+              zones: ["zone-1", "zone-2"]
+            - name: "group2"
+              zones: ["zone-3", "zone-4"]
+          - groups:
+            - zones: ["zone-5"]
+        fallback: # (6)
+          type: Any # (7)
+          exclude: # (8)
+            zones: ["zone-6"]
+      overprovisioningFactor: 120% # (9)
+```
+
+
+(1) In `inZone` section, you configure the priorities of tags used for load balancing. If list is empty or inZone is not specified, load balancing works normally, without taking any priorities into account. 
```

Comment:
this is a bit nit, but what do you think about `localZone`. It would be good to have terminology to name the zone in which you route the traffic. Also it would be good to explicitly say what is the value - it is a "tag" which is a base concept in Kuma.
```
localZone:
- tag: "k8s.io/node"
- tag: "k8s.io/az"
```

Reply:
localZone is for sure much better than `inZone` :)